### PR TITLE
Feat/471 cancel order

### LINKED ISSUE #471: Cancel order
As a user I want to be able to cancel an order.

- [x] Update vegawallet-service-api-client to include orderCancellation transaction in open api schema
- [x] Update to latest version of vegawallet-service-api-client
- [x] Add a cancel button to the order list
- [ ] Add a vega transaction dialog to prompt user to sign cancellation transaction in wallet
- [ ] Ensure order status is updated (This should already be working as the order subscription should already be running)

### PR DESCRIPTION
**Pull request by MadalinaRaicu**
# Related issues 🔗

Closes #471

# Description ℹ️

Add the ability to cancel orders. 

# Demo 📺
Screenshot 2022-06-24 at 21 54 09
Screenshot 2022-06-24 at 21 54 20
Screenshot 2022-06-24 at 21 56 09
Screenshot 2022-06-25 at 20 01 15

# Technical 👨‍🔧
- [x]  Update to latest version of vegawallet-service-api-client
- [x]  Add a cancel button to the order list
- [x]  Add a vega transaction dialog to prompt user to sign cancellation transaction in wallet
- [x]  Ensure order status is updated

### CODE REVIEW

File: libs/wallet/src/vega-transaction-dialog/vega-transaction-dialog.tsx [hunk 1]
```diff
@@ -5,18 +5,34 @@ import {
...
 
 interface OrderDialogProps {
   transaction: VegaTxState;
-  finalizedOrder: OrderEvent_busEvents_event_Order | null;
+  finalizedOrder: Order | null;
+  title?: string;
 }
 
-export const OrderDialog = ({
+export const VegaTransactionDialog = ({
```
Comment (dexturr) [inline libs/wallet/src/vega-transaction-dialog/vega-transaction-dialog.tsx:32]: chore (blocking): not all transactions are orders, so this should probably be VegaOrderTransactionDialog

File: libs/wallet/src/order-hooks/use-order-cancel.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+  const cancel = useCallback(
+    async (order) => {
+      if (!keypair) {
+        return;
+      }
+
+      setFinalizedOrder(null);
+
+      const res = await send({
```
Comment (dexturr) [inline libs/wallet/src/order-hooks/use-order-cancel.tsx:52]: chore (blocking): we need some form of error handling either here or where cancel is invoked. Probably here for reusability.

File: libs/wallet/src/order-hooks/use-order-cancel.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    variables: { partyId: keypair?.pub || '' },
+    skip: !id,
+    onSubscriptionData: ({ subscriptionData }) => {
+      if (!subscriptionData.data?.busEvents?.length) {
+        return;
+      }
+      // No types available for the subscription result
+      const matchingOrderEvent = subscriptionData.data.busEvents[0].event;
+
+      if (matchingOrderEvent && matchingOrderEvent.__typename === 'Order') {
```
Comment (dexturr) [inline libs/wallet/src/order-hooks/use-order-cancel.tsx:32]: thought (non-blocking): i wonder if it's worth generating type names as enums or keeping a map in VegaTypes to stop all these magic strings

File: libs/wallet/src/context.ts [hunk 4]
```diff
@@ -3,7 +3,7 @@ import type {
   TransactionResponse,
 } from '@vegaprotocol/vegawallet-service-api-client';
 import { createContext } from 'react';
-import type { VegaConnector } from './connectors';
+import type { VegaConnector } from './connectors/vega-connector';
```
Comment (dexturr) [inline libs/wallet/src/context.ts:6]: chore (blocking): this is still exported from the index, so i don't think this change is needed

File: libs/wallet/src/connectors/rest-connector.ts [hunk 5]
```diff
@@ -85,7 +85,7 @@ export class RestConnector implements VegaConnector {
     }
   }
 
-  async sendTx(body: TransactionSubmission) {
+  async sendTx(body?: TransactionSubmission) {
```
Comment (dexturr) [inline libs/wallet/src/connectors/rest-connector.ts:88]: question (blocking): why are we permitting empty transactions? These would immediately get rejected so I think this doesn't make and sense?

File: libs/order-list/src/setup-tests.ts [hunk 6]
```diff
@@ -1 +1,2 @@
 import '@testing-library/jest-dom';
+import 'jest-canvas-mock';
```
Comment (dexturr) [inline libs/order-list/src/setup-tests.ts:2]: question (blocking): why is this needed?

File: libs/wallet/src/order-hooks/order-event-query.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import { gql } from '@apollo/client';
+
+export const ORDER_EVENT_SUB = gql`
```
Comment (dexturr) [inline libs/wallet/src/order-hooks/order-event-query.ts:3]: nitpick (non-blocking): I would prefer this to be in the useOrderCancel file, as it is not shared.

**Review by dexturr — CHANGES_REQUESTED** [review]

File: libs/wallet/src/order-hooks/order-event-query.ts (hunk 7 above)
Comment (MadalinaRaicu) [reply libs/wallet/src/order-hooks/order-event-query.ts:3]: I would use this in useOrderEdit and maybe useOrderSubmit. eventually I would have all the order hooks living here

File: libs/wallet/src/vega-transaction-dialog/vega-transaction-dialog.tsx (hunk 1 above)
Comment (MadalinaRaicu) [reply libs/wallet/src/vega-transaction-dialog/vega-transaction-dialog.tsx]: updated in last commit

File: libs/order-list/src/setup-tests.ts (hunk 6 above)
Comment (MadalinaRaicu) [reply libs/order-list/src/setup-tests.ts]: can be removed, updated in last commit.

File: libs/wallet/src/order-hooks/use-order-cancel.tsx (hunk 3 above)
Comment (MadalinaRaicu) [reply libs/wallet/src/order-hooks/use-order-cancel.tsx]: i updated this to use BusEventType.Order

Comment (dexturr) [reply libs/wallet/src/order-hooks/use-order-cancel.tsx]: Can you undo that please. That is the name of the event type which is not nesseacrily the name of the entity in GQL. Leave as is, I just meant something I think we should look into, as this problem is across the whole codebase. Probably to use the introspection query to create an enum of type names or something but not something for now :)

Comment (MadalinaRaicu) [reply libs/wallet/src/order-hooks/use-order-cancel.tsx]: ok, alright.

File: libs/wallet/src/order-hooks/order-event-query.ts (hunk 7 above)
Comment (dexturr) [reply libs/wallet/src/order-hooks/order-event-query.ts:3]: We should not share queries, the idea behind GQL is to request what you need only, not everything. If we need to wait for an order and useEdit needs to check the price and useCancel does not these should be 2 separate queries. Sharing them gives us more headaches than it solves problems. Otherwise we are throwing away the majority of the value proposition of GQL. If the orders subs happen to for some reason need exactly the same field I guess we could consider sharing them but the default should always be separate queries unless you have a really good reason that they should be shared, e.g. trade list and charts do a whole bunch of data processing, which needs to not be duplicated for performance reasons, which is why those are shared.

Also chore (blocking): there are fields in this query which are unused can you please remove them. e.g. eventId, block.

File: libs/wallet/src/order-hooks/use-order-cancel.tsx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        setFinalizedOrder(matchingOrderEvent);
+      }
+    },
+  });
+
+  useEffect(() => {
+    if (finalizedOrder) {
+      resetTransaction();
```
Comment (dexturr) [inline libs/wallet/src/order-hooks/use-order-cancel.tsx:41]: question (blocking): @mattrussell36 I notice this is also present in the useSubmitOrder function so asking you haha, but this isn't cleaning up the subscription after the event has been fired. I know this is in a hook so once the element in unrendered the subscription will be cancelled but IMO we should lean up the listener here as well. Was there a reason we decided not to do this in useOrderSubmit?

It's not the worst thing in the world as this is filtered by partyId but wanted to double check. I just worry if this is used as part of non-transient UI we could end up with a memory leek, albeit a slow one

File: libs/wallet/src/order-hooks/use-order-cancel.tsx [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+          },
+        });
+
+        if (res?.signature) {
+          setId(determineId(res.signature));
+        }
+        return res;
+      } catch (e) {
+        Sentry.captureException(e);
+        return;
```
Comment (dexturr) [inline libs/wallet/src/order-hooks/use-order-cancel.tsx:69]:
```suggestion
```

File: libs/order-list/src/lib/components/cancel-order-dialog/cancel-dialog.tsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+export const CancelDialog = ({
+  orderDialogOpen,
+  setOrderDialogOpen,
+  finalizedOrder,
+  transaction,
+  reset,
+}: CancelDialogProps) => {
+  const getDialogIntent = () => {
+    if (finalizedOrder) {
```
Comment (edd) [inline libs/order-list/src/lib/components/cancel-order-dialog/cancel-dialog.tsx:23]: **nitpick (non-blocking)**: I found this if logic a little hard to parse because `finalizedOrder` implied to me that we were *placing* an Order, whereas obviously we're waiting for the *updated* order to come in and to confirm that it's actually cancelled.

It's not worth making a change for this.

**Review by edd — APPROVED** [review]
:+1: One trivial variable naming comment, and I think this could have been broken up to better contain just the Order Cancellation stuff, but at least one of the split out PRs would have been pretty large anyway.

File: libs/wallet/src/order-hooks/use-order-cancel.tsx (hunk 9 above)
Comment (MadalinaRaicu) [reply libs/wallet/src/order-hooks/use-order-cancel.tsx]: this won't build cause not all code paths return a value in the callback. I need to have a return on this catch.

File: libs/order-list/src/lib/components/cancel-order-dialog/cancel-dialog.tsx (hunk 10 above)
Comment (MadalinaRaicu) [reply libs/order-list/src/lib/components/cancel-order-dialog/cancel-dialog.tsx:23]: updatedOrder might be better

File: libs/order-list/src/lib/components/order-list-manager/order-list-manager.tsx [hunk 11]
```diff
@@ -39,17 +38,17 @@ export const OrderListManager = ({ partyId }: OrderListManagerProps) => {
       const rowNode = gridRef.current.api.getRowNode(d.id);
 
       if (rowNode) {
-        if (!isEqual) {
-          update.push(d);
+        if (!isEqual(d, rowNode.data)) {
```
Comment (mattrussell36) [inline libs/order-list/src/lib/components/order-list-manager/order-list-manager.tsx:41]: comment (non-blocking): Thanks for fixing this one

File: libs/order-list/src/lib/components/order-list/order-list.tsx [hunk 12]
```diff
@@ -1,24 +1,78 @@
...
 
 export const OrderList = forwardRef<AgGridReact, OrderListProps>(
   ({ data }, ref) => {
+    const [orderDialogOpen, setOrderDialogOpen] = useState(false);
+    const { transaction, finalizedOrder, reset, cancel } = useOrderCancel();
+    return (
+      <>
+        <OrderListTable data={data} cancel={cancel} ref={ref} />
+        <CancelDialog
+          {...{
```
Comment (mattrussell36) [inline libs/order-list/src/lib/components/order-list/order-list.tsx:37]: chore (blocking): Unnecessary spread

File: libs/order-list/src/lib/components/order-list/order-list.tsx [hunk 13]
```diff
@@ -27,20 +81,19 @@ export const OrderList = forwardRef<AgGridReact, OrderListProps>(
         defaultColDef={{ flex: 1, resizable: true }}
         style={{ width: '100%', height: '100%' }}
         getRowId={({ data }) => data.id}
+        rowHeight={40}
       >
         <AgGridColumn
-          headerName={t('Market')}
+          headerName="Market"
```
Comment (mattrussell36) [inline libs/order-list/src/lib/components/order-list/order-list.tsx:87]: chore (blocking): Can we keep the `t` function?

File: libs/order-list/src/lib/components/order-list/order-list.tsx [hunk 14]
```diff
@@ -27,20 +81,19 @@ export const OrderList = forwardRef<AgGridReact, OrderListProps>(
...
         />
         <AgGridColumn
-          headerName={t('Amount')}
+          headerName="Amount"
           field="size"
           cellClass="font-mono"
           valueFormatter={({ value, data }: ValueFormatterParams) => {
             const prefix = data.side === Side.Buy ? '+' : '-';
-            return (
-              prefix + addDecimal(value, data.market.positionDecimalPlaces)
```
Comment (mattrussell36) [inline libs/order-list/src/lib/components/order-list/order-list.tsx:42]: chore (blocking): Why are you removing this? Amount still needs formatting with DPs due to fractional orders

File: libs/order-list/src/lib/components/order-list/order-list.tsx [hunk 15]
```diff
@@ -1,24 +1,78 @@
...
+          }}
+        />
+      </>
+    );
+  }
+);
+
+export const OrderListTable = forwardRef<AgGridReact, OrderListTableProps>(
+  ({ data, cancel }, ref) => {
+    const CancelRendererButton = ({ data }: ICellRendererParams) => {
```
Comment (mattrussell36) [inline libs/order-list/src/lib/components/order-list/order-list.tsx:52]: chore (blocking): I don't see benefit of creating a function here its not used mored than once and just results in you having to scroll up again once you get to this cell in the render logic below.  Can we move it inline?
```
cellRenderer={({ data} : ICellRendererPrarams) => ...}
```

File: libs/wallet/src/order-hooks/use-order-cancel.tsx (hunk 8 above)
Comment (mattrussell36) [reply libs/wallet/src/order-hooks/use-order-cancel.tsx:41]: chore (blocking): Additionally, not sure why I used a hook to reset transaction here it just makes things more confusing. Can we move resetTransaction() to just be called immediately after setUpdatedOrder. The code is more readable that way. Also could you make the same change in the use-order-submit hook.

File: libs/wallet/src/order-hooks/use-order-edit.tsx [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+  OrderEvent,
+  OrderEventVariables,
+  OrderEvent_busEvents_event_Order,
+} from './__generated__/OrderEvent';
+import { ORDER_EVENT_SUB } from './order-event-query';
+
+export const useOrderEdit = () => {
+  const { keypair } = useVegaWallet();
+  const { send, transaction, reset: resetTransaction } = useVegaTransaction();
+  const [finalizedOrder, setFinalizedOrder] =
```
Comment (mattrussell36) [inline libs/wallet/src/order-hooks/use-order-edit.tsx:17]: suggestion(non blocking):
```suggestion
  const [editedOrder, setEditedOrder] =
```

File: libs/wallet/src/order-hooks/use-order-edit.tsx [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+      // No types available for the subscription result
+      const matchingOrderEvent = subscriptionData.data.busEvents[0].event;
+
+      if (matchingOrderEvent && matchingOrderEvent.__typename === 'Order') {
+        setFinalizedOrder(matchingOrderEvent);
+      }
+    },
+  });
+
+  useEffect(() => {
```
Comment (mattrussell36) [inline libs/wallet/src/order-hooks/use-order-edit.tsx:38]: chore (blocking): As per my other comments. Lets remove the useEffect and just call this underneath `setFinalizedOrder` above

File: libs/wallet/src/provider.tsx [hunk 18]
```diff
@@ -104,7 +104,7 @@ export const VegaWalletProvider = ({ children }: VegaWalletProviderProps) => {
       disconnect,
       connector: connector.current,
       sendTx,
-    };
+    } as VegaWalletContextShape;
```
Comment (mattrussell36) [inline libs/wallet/src/provider.tsx:107]: question (blocking): Why the type cast here?

Comment (MadalinaRaicu) [reply libs/wallet/src/provider.tsx:107]: formatting error

File: libs/order-list/src/lib/components/cancel-order-dialog/cancel-dialog.tsx [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+
+interface CancelDialogProps {
+  orderDialogOpen: boolean;
+  setOrderDialogOpen: (isOpen: boolean) => void;
+  finalizedOrder: Order | null;
+  transaction: VegaTxState;
+  reset: () => void;
+}
+
+export const CancelDialog = ({
```
[the remark this answers is not in the document]
Comment (MadalinaRaicu) [reply libs/order-list/src/lib/components/cancel-order-dialog/cancel-dialog.tsx:15]: I will look into refactoring this in the #463 ticket

File: libs/wallet/src/order-hooks/use-order-cancel.tsx [hunk 20]
```diff
@@ -0,0 +1,88 @@
+import { useCallback, useEffect, useState } from 'react';
```
[the remark this answers is not in the document]
Comment (MadalinaRaicu) [reply libs/wallet/src/order-hooks/use-order-cancel.tsx:1]: interesting, I was trying to fit everything in wallet lib. i would love to do this in #463

File: libs/wallet/src/provider.tsx (hunk 18 above)
Comment (mattrussell36) [reply libs/wallet/src/provider.tsx:107]: Hm there's a larger problem then. I have a feeling its due to sendTx types?

File: libs/wallet/src/order-hooks/use-order-cancel.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (mattrussell36) [reply libs/wallet/src/order-hooks/use-order-cancel.tsx:1]: Lets make a different ticket for this 👍

File: libs/wallet/src/order-hooks/use-order-edit.tsx (hunk 16 above)
Comment (MadalinaRaicu) [reply libs/wallet/src/order-hooks/use-order-edit.tsx:17]: this needs to come in another PR -  i will remove it for now

File: libs/deal-ticket/src/components/deal-ticket-manager.tsx [hunk 21]
```diff
@@ -49,12 +48,12 @@ export const DealTicketManager = ({
   };
 
   useEffect(() => {
-    if (transaction.status !== VegaTxStatus.Default) {
+    if (transaction.status !== VegaTxStatus.Default || finalizedOrder) {
```
[the remark this answers is not in the document]
Comment (MadalinaRaicu) [reply libs/deal-ticket/src/components/deal-ticket-manager.tsx:51]: I added this because the `Order placed` modal would not appear otherwise

[the remark this answers is not in the document]
Comment (MadalinaRaicu) [reply libs/deal-ticket/src/components/deal-ticket-manager.tsx:51]: It would get stuck into `Awaiting network confirmation` as the tx status comes very quickly into Default. I tried to quickly place an additional order and it works fine.

[the remark this answers is not in the document]
Comment (MadalinaRaicu) [reply libs/deal-ticket/src/components/deal-ticket-manager.tsx:51]: It is actually the same dialog that reopens. 
As per your suggestion, I will refactor this dialog to be used for both submission and cancellation (without any dialog wrappers).
I think I should update this in a separate PR.

File: libs/wallet/src/order-hooks/use-order-cancel.tsx (hunk 8 above)
Comment (MadalinaRaicu) [reply libs/wallet/src/order-hooks/use-order-cancel.tsx]: Updated to use resetTransaction immediately after setUpdatedOrder. 
Let me use sth else instead of useSubscription in another PR. I will refactor a bit order-list and move all the order hooks from the wallet.

Comment (MadalinaRaicu) [reply libs/wallet/src/order-hooks/use-order-cancel.tsx]: https://github.com/vegaprotocol/frontend-monorepo/issues/657

**Review by mattrussell36 — CHANGES_REQUESTED** [review]

File: libs/order-list/src/lib/components/mocks/generate-orders.ts [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+
+export const marketOrder = generateOrder({
+  id: 'market-order',
+  type: OrderType.Market,
+  status: OrderStatus.Active,
+});
+
+export const generateMockOrders = (): Orders_party_orders[] => {
+  return [
+    {
```
[the remark this answers is not in the document]
Comment (MadalinaRaicu) [reply libs/order-list/src/lib/components/mocks/generate-orders.ts:75]: If I import this from the order-list it messes up with my cypress tests. 
image

File: libs/order-list/src/lib/components/cancel-order-dialog/cancel-dialog.spec.tsx [hunk 23]
```diff
@@ -0,0 +1,71 @@
+import { render, screen } from '@testing-library/react';
+import { OrderStatus, OrderType } from '@vegaprotocol/types';
+import type { VegaTxState, Order } from '@vegaprotocol/wallet';
+import { VegaTxStatus } from '@vegaprotocol/wallet';
+import { CancelDialog } from './cancel-dialog';
+
+describe('CancelDialog', () => {
+  it('should render when an order is successfully cancelled', () => {
+    const orderDialogOpen = true;
+    const setOrderDialogOpen = () => false;
```
[the remark this answers is not in the document]
Comment (MadalinaRaicu) [reply libs/order-list/src/lib/components/cancel-order-dialog/cancel-dialog.spec.tsx]: updated in last commit

**Review by mattrussell36 — APPROVED** [review]

File: libs/order-list/src/lib/components/mocks/generate-orders.ts [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+
+export const marketOrder = generateOrder({
+  id: 'market-order',
+  type: OrderType.Market,
+  status: OrderStatus.Active,
+});
+
+export const generateMockOrders = (): Orders_party_orders[] => {
+  return [
+    generateOrder({
```
Comment (dexturr) [inline libs/order-list/src/lib/components/mocks/generate-orders.ts:76]: nitpick (non-blocking): lot of duplicated fields here, might be nice to have a default object and spread that into these

File: libs/wallet/src/order-hooks/order-event-query.ts [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+  subscription OrderEvent($partyId: ID!) {
+    busEvents(partyId: $partyId, batchSize: 0, types: [Order]) {
+      eventId
+      block
```
Comment (dexturr) [inline libs/wallet/src/order-hooks/order-event-query.ts:7]:
```suggestion
```

**Review by dexturr — APPROVED** [review]
